### ISSUE TITLE
PySide2/PyQt5: Graphics issues in QScrollArea for OSX

### ISSUE THREAD

**Issue by pshriwise** [matplotlib/matplotlib#14160]
### Bug report

**Bug summary**

Figure graphics aren't re-drawn correctly in a QScrollArea when using `matplotlib>2.2.4` on OSX. I've put together a short program based on one of the examples which demonstrates this.

This seems specific to Macs as I've run the same code on Ubuntu without these issues.

**Code for reproduction**

```python
import sys
import os
import random
import numpy as np

# PySide2 here gives the same result
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

import matplotlib
# Make sure that we are using QT5
matplotlib.use('Qt5Agg')

from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg as FigureCanvas
from matplotlib.figure import Figure

class MyMplCanvas(FigureCanvas):
    """Ultimately, this is a QWidget (as well as a FigureCanvasAgg, etc.)."""

    def __init__(self, parent=None, width=5, height=4, dpi=100):
        fig = Figure(figsize=(width, height), dpi=dpi)
        self.axes = fig.add_subplot(111)

        self.compute_initial_figure()

        FigureCanvas.__init__(self, fig)
        self.setParent(parent)

        FigureCanvas.setSizePolicy(self,
                                   QSizePolicy.Expanding,
                                   QSizePolicy.Expanding)
        FigureCanvas.updateGeometry(self)

    def compute_initial_figure(self):
        pass

class MyStaticMplCanvas(MyMplCanvas):
    """Simple canvas with a sine plot."""

    def compute_initial_figure(self):
        t = np.linspace(0.0, 3.0, 2000)
        s = np.sin(2*np.pi*t)
        self.axes.plot(t, s)

class Widget(QWidget):

    def __init__(self, parent= None):
        super(Widget, self).__init__()

        #Container Widget
        widget = QWidget()
        #Layout of Container Widget
        layout = QVBoxLayout(self)

        fc = MyStaticMplCanvas()
        layout.addWidget(fc)
        widget.setLayout(layout)

        #Scroll Area Properties
        scroll = QScrollArea()
        scroll.setVerticalScrollBarPolicy(Qt.ScrollBarAlwaysOn)
        scroll.setHorizontalScrollBarPolicy(Qt.ScrollBarAlwaysOn)
        scroll.setWidgetResizable(False)
        scroll.setWidget(widget)

        #Scroll Area Layer add
        vLayout = QVBoxLayout(self)
        vLayout.addWidget(scroll)
        self.setLayout(vLayout)

if __name__ == '__main__':
    app = QApplication(sys.argv)
    dialog = Widget()
    dialog.show()
    app.exec_()

```

The plot looks like this with window larger than the plot:

<img width="682" alt="Screen Shot 2019-05-07 at 1 09 30 PM" src="https://user-images.githubusercontent.com/4563941/57323314-2d2b5f80-70cb-11e9-84d6-27e49051e8ae.png">

**Actual outcome**

After resizing the window and scrolling to see the rest of the canvs, parts of the plot aren't drawn correctly both during and after the scroll event:

<img width="473" alt="Screen Shot 2019-05-07 at 1 10 29 PM" src="https://user-images.githubusercontent.com/4563941/57323342-3ddbd580-70cb-11e9-93b1-ad722ea98bd1.png">

This is true for any version of matplotlib>=2.2.4.

**Expected outcome**

When reverting the matplotlib version to 2.2.4 using pip3, the plot appears normally during and after scrolling:

<img width="485" alt="Screen Shot 2019-05-07 at 1 12 14 PM" src="https://user-images.githubusercontent.com/4563941/57323384-50eea580-70cb-11e9-9148-1334daf244d1.png">

**Matplotlib version**

Matplotlib version >= 2.2.4 installed with pip3.
Python 3.7
OSX 10.14.3
Backend: backend_qt5agg
Other libraries (installed with pip3)
  - PySide2 (5.12.3)
  - PyQt5.QtCore.QT_VERSION_STR (5.12.2)
  - PyQt5.QtCore.PYQT_VERSION_STR (5.12.2)

**Comment by pshriwise** [comment]
I created a fix for this problem [here](https://github.com/matplotlib/matplotlib/compare/master...pshriwise:qt5_bbox_fix?expand=1). Whether or not it is a robust fix for all cases, I'm not sure. 

I found that the root of the problem is the rectangle of the `QPaintEvent` has an origin and width that does not represent the size of the canvas, but rather the currently visible area of the `FigureCanvas`, causing only part of the plot area to be re-drawn.

This is more intuitive when looking at a `QRubberBand` in the `QScrollArea` where a similar graphics problem occurs. The `QPaintEvent` rectangle in this case represents the selected area rather than the size of the `FigureCanvas`. 



Before fix:

![image (1)](https://user-images.githubusercontent.com/4563941/58223779-ac7e7d00-7ce0-11e9-9182-afe1c901eace.png)

Where the bounding box is: 
`Bbox(x0=141.0, y0=575.0, x1=797.0, y1=1003.0)`

After fix:

![image](https://user-images.githubusercontent.com/4563941/58223783-b0120400-7ce0-11e9-9d6a-fa8654de1163.png)

`Bbox(x0=0.0, y0=0.0, x1=1352.0, y1=1198.0)`

So. I don't know if this is the correct fix for this issue, but maybe it will shed some light on the problem and lead to a proper fix in the near future?
